rpcs3/Emu/Cell/PPUThread.cpp

```diff
@@ -1654,24 +1682,53 @@ static bool ppu_store_reservation(ppu_thread& ppu, u32 addr, u64 reg_value)
 			default: break;
 			}
 
-			if (res == rtime && vm::reservation_trylock(res, rtime))
+			res += 1;
+
+			if (data.compare_and_swap_test(old_data, reg_value))
 			{
-				const bool ret = data.compare_and_swap_test(old_data, reg_value);
-				res.release(rtime + 128);
-				return ret;
+				res += 127;
+				return true;
 			}
 
+			res -= 1;
 			return false;
 		}
 
-		if (!vm::reservation_trylock(res, rtime))
+		while (true)
 		{
-			return false;
+			auto [_old, _ok] = res.fetch_op([&](u64& r)
+			{
+				if (r & vm::rsrv_unique_lock)
+				{
+					return false;
+				}
+
+				r += 1;
+				return true;
+			});
+
+			// Give up if reservation has been updated
+			if ((_old & -128) != rtime)
+			{
+				return false;
+			}
+
+			if (_ok)
+			{
+				break;
+			}
+
+			busy_wait(100);
```

Comment:
Now that its not a single try to lock but a loop cpu_thread::check_state() must be called here. Previously it was doing vm::passive_unlock/lock for it but check_state is more optimized.